### DISCUSSION TITLE
Provide schema for command args

### DISCUSSION THREAD

**Discussion by akbyrd** [microsoft/vscode-discussions#541] [Extension Development QnA] [answered]
Some built-in commands such as `workbench.action.tasks.runTask` provide a schema for their args, which enables auto-completion when creating a keybinding to invoke that command.

https://github.com/microsoft/vscode/blob/0d5e948d863153807cfe07ecd20dc10374eea6d2/src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts#L421-L446

![image](https://user-images.githubusercontent.com/484784/227801050-36cf771b-684b-497c-8ec9-356fe59ab650.png)

Can extensions also provide an args schema for their contributed commands? If so, how?

**Comment by isidorn** [comment]
Good question, but I think this is not possible.
@aeschli can confirm

**Reply by akbyrd** [reply]
I don't know if it's the "correct" way, but I did notice that I can use the `jsonValidation` contribution point to create validation in keybindings.json files. This provides warnings when the args are not formatted correctly and also provides both auto-completion and snippets.

Unfortunately it doesn't work in some situations, such as command input variables in tasks/launch files, but I don't think the built-in schemas work there either.

I don't know if typescript supports introspection, but this does have me idly wondering if the schema could be inferred from the arguments of the function passed to `registerCommand`.

```
"contributes": {
  "jsonValidation": [
    {
      "fileMatch": "%APP_SETTINGS_HOME%/keybindings.json",
      "url": "./src/commands.schema.jsonc"
    },
    {
      "fileMatch": "%APP_SETTINGS_HOME%/profiles/*/keybindings.json",
      "url": "./src/commands.schema.jsonc"
    }
  ],
}
```

```
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "type": "array",
  "items": {
    "type": "object",
    "if": {
      "required": [ "command" ],
      "properties": { "command": { "const": "akbyrd.task.runWithArgs" } }
    },
    "then": {
      "required": [ "args" ],
      "properties": {
        "args": {
          "type": [ "string", "object" ]
        }
      },
      "if": {
        "properties": { "args": { "type": "object" } }
      },
      "then": {
        "properties": {
          "args": {
            "required": [ "task" ],
            "properties": {
              "task": {
                "description": "The standard arguments to workbench.action.tasks.runTask",
                "type": [ "string", "object" ]
              },
              "taskArgs": {
                "description": "An object filled task arguments. Keys are argument names and values are argument values.",
                "type": "object",
                "patternProperties": {
                  ".*": { "not": { "type": [ "object", "array" ] } }
                }
              }
            },
            "defaultSnippets": [
              { "bodyText": "{ \"task\": \"${1:taskName}\", \"taskArgs\": { \"${4:argName}\": \"${5:argValue}\" } }" }
            ],
            "if": {
              "properties": { "task": { "type": "object"} }
            },
            "then": {
              "properties": {
                "task": {
                  "properties": {
                    "task": { "type": "string" },
                    "type": { "type": "string" }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by aeschli** [comment, marked as answer]
It's something we should add to the command contribution point. Please file a feature request.
Adding as an additional schema to `%APP_SETTINGS_HOME%/keybindings.json` is too cumbersome and not recommended.
